e2xgrader/utils/mode.py

```diff
@@ -0,0 +1,128 @@
+import os
+from typing import Dict
+
+from jupyter_core.paths import jupyter_config_path
+
+from .. import _jupyter_server_extension_paths
+from ..extensions.utils import discover_nbextensions, get_notebook_config_manager
+
+
+def get_nbextension_config() -> Dict[str, Dict[str, bool]]:
+    """
+    Get the configuration for nbextensions.
+
+    Returns:
+        A dictionary containing the configuration for nbextensions.
+        The dictionary has the following structure:
+        {
+            'tree': {
+                'nbextension_name': True/False,
+                ...
+            },
+            'notebook': {
+                'nbextension_name': True/False,
+                ...
+            }
+        }
+    """
+    config = dict(tree=dict(), notebook=dict())
+    for path in jupyter_config_path():
+        config_path = os.path.join(path, "nbconfig")
+        contextmanager = get_notebook_config_manager()(config_dir=config_path)
+        for key in config:
+            config[key].update(contextmanager.get(key))
+    return config
+
+
+def get_serverextension_config() -> Dict[str, bool]:
+    """
+    Retrieves the server extension configuration from the Jupyter notebook configuration files.
+
+    Returns:
+        A dictionary containing the server extension configuration, where the keys are the
+        extension names and the values are boolean values indicating whether the extension
+        is enabled or not.
+    """
+    config = dict()
+    for path in jupyter_config_path():
+        contextmanager = get_notebook_config_manager()(config_dir=path)
+        config.update(contextmanager.get("jupyter_notebook_config"))
+    return config.get("NotebookApp", dict()).get("nbserver_extensions", dict())
+
+
+def infer_nbextension_mode() -> str:
+    """
+    Infer the active mode for the nbextension based on the current configuration.
+
+    Returns:
+        str: The active mode for the nbextension.
+
+    Raises:
+        ValueError: If more than one mode is active or if tree and notebook extensions don't match.
+    """
+    config = get_nbextension_config()
+    modes = ["teacher", "student_exam", "student"]
+    is_active = dict(tree=[], notebook=[])
+    for mode in modes:
```

Comment:
not sure what's happening here and why... could use a quick comment...

Reply:
We infer which extensions are active in the tree section and which in the notebook section. Then for each mode we check if the respective tree and notebook extensions are active. Finally we check if only the extensions for one mode are active and not the extensions for multiple modes.